jolt-core/src/subprotocols/grand_product.rs

```diff
@@ -1,530 +1,1699 @@
-        let mut rand: Vec<F> = Vec::new();
-        let num_layers = self.proof.len();
-
-        let mut claims_to_verify = claims_prod_vec.to_owned();
-        for (num_rounds, i) in (0..num_layers).enumerate() {
-            // produce random coefficients, one for each instance
-            let coeff_vec =
-                transcript.challenge_vector::<F>(b"rand_coeffs_next_layer", claims_to_verify.len());
+    fn num_layers(&self) -> usize {
+        self.layers.len()
+    }
 
-            // produce a joint claim
-            let claim = (0..claims_to_verify.len())
-                .map(|i| claims_to_verify[i] * coeff_vec[i])
-                .sum();
+    fn claims(&self) -> Vec<F> {
+        let last_layers = &self.layers[self.num_layers() - 1];
+        assert_eq!(last_layers.layer_len, 2);
+        last_layers
+            .layers
+            .iter()
+            .map(|layer| layer[0] * layer[1])
+            .collect()
+    }
 
-            let (claim_last, rand_prod) = self.proof[i].verify(claim, num_rounds, 3, transcript);
+    fn layers(&'_ mut self) -> impl Iterator<Item = &'_ mut dyn BatchedGrandProductLayer<F>> {
+        self.layers
+            .iter_mut()
+            .map(|layer| layer as &mut dyn BatchedGrandProductLayer<F>)
+            .rev()
+    }
+}
 
-            let claims_prod_left = &self.proof[i].claims_poly_A;
-            let claims_prod_right = &self.proof[i].claims_poly_B;
-            assert_eq!(claims_prod_left.len(), claims_prod_vec.len());
-            assert_eq!(claims_prod_right.len(), claims_prod_vec.len());
+/// Represents a single layer of a single grand product circuit using a sparse vector,
+/// i.e. a vector containing (index, value) pairs.
+/// Nodes with value 1 are omitted from the sparse vector.
+/// Like `DenseGrandProductLayer`, a `SparseGrandProductLayer` is assumed to be
+/// arranged in "interleaved" order:
+///      Λ        Λ        Λ        Λ
+///     / \      / \      / \      / \
+///   L0   1   L1   R1   1   1   L3   1   <- This is layer would be represented as [(0, L0), (2, L1), (3, R1), (6, L3)]
+pub type SparseGrandProductLayer<F> = Vec<(usize, F)>;
+
+/// A "dynamic density" grand product layer can switch from sparse representation
+/// to dense representation once it's no longer sparse (after binding).
+#[derive(Debug, Clone, PartialEq)]
+pub enum DynamicDensityGrandProductLayer<F: JoltField> {
+    Sparse(SparseGrandProductLayer<F>),
+    Dense(DenseGrandProductLayer<F>),
+}
 
-            for i in 0..claims_prod_vec.len() {
-                transcript.append_scalar(b"claim_prod_left", &claims_prod_left[i]);
-                transcript.append_scalar(b"claim_prod_right", &claims_prod_right[i]);
+/// The threshold at which a `DynamicDensityGrandProductLayer` switches from sparse to
+/// dense representation. If the layer has >DENSIFICATION_THRESHOLD fraction of non-1
+/// values, it'll switch to the dense representation.
+const DENSIFICATION_THRESHOLD: f64 = 0.8;
```

Comment:
This is used to switch both at the Layer level and also the Tree level. Both in `DynamicDensityGrandProductLayer::layer_output` and `BatchedSparseGrandProductLayer::bind`. Suggested update:

The threshold at which a `DynamicDensityGrandProductLayer` switches from sparse to
dense representation for the parent layer and `BatchedSparseGrandProductLayer` switches from sparse to dense for its polynomial representation. If the layer has >DENSIFICATION_THRESHOLD fraction of non-1
values, it'll switch to the dense representation. Value tuned experimentally.

Reply:
done